File: internal/cliclient/projects.go
package cliclient

import (
	"context"
	"fmt"
	"net/url"
)

// ListProjects returns all projects.
func (c *Client) ListProjects(ctx context.Context) ([]Project, error) {
	var projects []Project
	_, err := c.Get(ctx, "/projects", &projects)
	if err != nil {
		return nil, err
	}
	return projects, nil
}

// GetProject returns a project by ID.
func (c *Client) GetProject(ctx context.Context, id string) (*Project, error) {
	var project Project
	_, err := c.Get(ctx, fmt.Sprintf("/projects/%s", id), &project)
	if err != nil {
		return nil, err
	}
	return &project, nil
}

// CreateProject creates a new project.
func (c *Client) CreateProject(ctx context.Context, req CreateProjectRequest) (*Project, error) {
	var project Project
	_, err := c.Post(ctx, "/projects", req, &project)
	if err != nil {
		return nil, err
	}
	return &project, nil
}

// DeleteProject deletes a project by ID.
func (c *Client) DeleteProject(ctx context.Context, id string) error {
	_, err := c.Delete(ctx, fmt.Sprintf("/projects/%s", id))
	return err
}

// GetProjectPackages returns packages for a project.
func (c *Client) GetProjectPackages(ctx context.Context, projectID string) ([]Package, error) {
	var pkgs []Package
	_, err := c.Get(ctx, fmt.Sprintf("/projects/%s/packages", projectID), &pkgs)
	if err != nil {
		return nil, err
	}
	return pkgs, nil
}

// GetProjectPublications returns publications for a project.
func (c *Client) GetProjectPublications(ctx context.Context, projectID string) ([]Publication, error) {
	var pubs []Publication
	_, err := c.Get(ctx, fmt.Sprintf("/projects/%s/publications", projectID), &pubs)
	if err != nil {
		return nil, err
	}
	return pubs, nil
}

// GetPublishDefaults returns suggested defaults for publishing a project.
func (c *Client) GetPublishDefaults(ctx context.Context, projectID string, registryID ...string) (*PublishDefaults, error) {
	var defaults PublishDefaults
	path := fmt.Sprintf("/projects/%s/publish-defaults", projectID)
	if len(registryID) > 0 && registryID[0] != "" {
		path += "?registry_id=" + url.QueryEscape(registryID[0])
	}
	_, err := c.Get(ctx, path, &defaults)
	if err != nil {
		return nil, err
	}
	return &defaults, nil
}

// GetProjectVersions returns versions for a project.
func (c *Client) GetProjectVersions(ctx context.Context, projectID string) ([]ProjectVersion, error) {
	var versions []ProjectVersion
	_, err := c.Get(ctx, fmt.Sprintf("/projects/%s/versions", projectID), &versions)
	if err != nil {
		return nil, err
	}
	return versions, nil
}

// GetVersionPixiToml returns the pixi.toml for a specific version.
func (c *Client) GetVersionPixiToml(ctx context.Context, projectID string, version int32) (string, error) {
	content, _, err := c.GetText(ctx, fmt.Sprintf("/projects/%s/versions/%d/pixi-toml", projectID, version))
	if err != nil {
		return "", err
	}
	return content, nil
}

// GetVersionPixiLock returns the pixi.lock for a specific version.
func (c *Client) GetVersionPixiLock(ctx context.Context, projectID string, version int32) (string, error) {
	content, _, err := c.GetText(ctx, fmt.Sprintf("/projects/%s/versions/%d/pixi-lock", projectID, version))
	if err != nil {
		return "", err
	}
	return content, nil
}

// GetProjectTags returns server-side tags for a project.
func (c *Client) GetProjectTags(ctx context.Context, projectID string) ([]ProjectTag, error) {
	var tags []ProjectTag
	_, err := c.Get(ctx, fmt.Sprintf("/projects/%s/tags", projectID), &tags)
	if err != nil {
		return nil, err
	}
	return tags, nil
}

// PushVersion pushes a new version to the server with a tag.
func (c *Client) PushVersion(ctx context.Context, projectID string, req PushRequest) (*PushResponse, error) {
	var resp PushResponse
	_, err := c.Post(ctx, fmt.Sprintf("/projects/%s/push", projectID), req, &resp)
	if err != nil {
		return nil, err
	}
	return &resp, nil
}

// PublishProject publishes a project to a registry.
func (c *Client) PublishProject(ctx context.Context, projectID string, req PublishRequest) (*PublishResponse, error) {
	var resp PublishResponse
	_, err := c.Post(ctx, fmt.Sprintf("/projects/%s/publish", projectID), req, &resp)
	if err != nil {
		return nil, err
	}
	return &resp, nil
}

// InstallProject queues an environment install for a project.
// Returns the queued Job (install runs asynchronously on the server).
func (c *Client) InstallProject(ctx context.Context, projectID string) (*Job, error) {
	var job Job
	_, err := c.Post(ctx, fmt.Sprintf("/projects/%s/install", projectID), nil, &job)
	if err != nil {
		return nil, err
	}
	return &job, nil
}

// UninstallProject queues removal of a project's installed environment.
func (c *Client) UninstallProject(ctx context.Context, projectID string) (*Job, error) {
	var job Job
	_, err := c.Post(ctx, fmt.Sprintf("/projects/%s/uninstall", projectID), nil, &job)
	if err != nil {
		return nil, err
	}
	return &job, nil
}

// RollbackProject queues a server-side rollback to a previous version.
// Returns the queued Job (rollback runs asynchronously on the server).
func (c *Client) RollbackProject(ctx context.Context, projectID string, versionNumber int) (*Job, error) {
	req := RollbackRequest{VersionNumber: versionNumber}
	var job Job
	_, err := c.Post(ctx, fmt.Sprintf("/projects/%s/rollback", projectID), req, &job)
	if err != nil {
		return nil, err
	}
	return &job, nil
}

File: frontend/src/api/projects.ts
import type {
  CreateProjectRequest,
  Job,
  Project,
  ProjectTag,
  ProjectVersion,
  RollbackRequest,
} from '@/types';
import { apiClient } from './client';

export const projectsApi = {
  list: async (): Promise<Project[]> => {
    const { data } = await apiClient.get('/projects');
    return data;
  },

  get: async (id: string): Promise<Project> => {
    const { data } = await apiClient.get(`/projects/${id}`);
    return data;
  },

  create: async (req: CreateProjectRequest): Promise<Project> => {
    const { data } = await apiClient.post('/projects', req);
    return data;
  },

  delete: async (id: string): Promise<void> => {
    await apiClient.delete(`/projects/${id}`);
  },

  getPixiToml: async (id: string): Promise<{ content: string }> => {
    const { data } = await apiClient.get(`/projects/${id}/pixi-toml`);
    return data;
  },

  // Version management
  listVersions: async (id: string): Promise<ProjectVersion[]> => {
    const { data } = await apiClient.get(`/projects/${id}/versions`);
    return data;
  },

  getVersion: async (
    id: string,
    versionNumber: number,
  ): Promise<ProjectVersion> => {
    const { data } = await apiClient.get(
      `/projects/${id}/versions/${versionNumber}`,
    );
    return data;
  },

  downloadLockFile: async (
    id: string,
    versionNumber: number,
  ): Promise<string> => {
    const { data } = await apiClient.get(
      `/projects/${id}/versions/${versionNumber}/pixi-lock`,
      {
        responseType: 'text',
      },
    );
    return data;
  },

  downloadManifest: async (
    id: string,
    versionNumber: number,
  ): Promise<string> => {
    const { data } = await apiClient.get(
      `/projects/${id}/versions/${versionNumber}/pixi-toml`,
      {
        responseType: 'text',
      },
    );
    return data;
  },

  rollback: async (id: string, req: RollbackRequest): Promise<Job> => {
    const { data } = await apiClient.post(`/projects/${id}/rollback`, req);
    return data;
  },

  listTags: async (id: string): Promise<ProjectTag[]> => {
    const { data } = await apiClient.get(`/projects/${id}/tags`);
    return data;
  },

  savePixiToml: async (id: string, content: string): Promise<void> => {
    await apiClient.put(`/projects/${id}/pixi-toml`, { content });
  },

  solveProject: async (id: string): Promise<Job> => {
    const { data } = await apiClient.post(`/projects/${id}/solve`);
    return data;
  },

  install: async (id: string): Promise<Job> => {
    const { data } = await apiClient.post(`/projects/${id}/install`);
    return data;
  },

  uninstall: async (id: string): Promise<Job> => {
    const { data } = await apiClient.post(`/projects/${id}/uninstall`);
    return data;
  },
};
